azurerm/_modules/aks/node_pool/main.tf

```diff
@@ -0,0 +1,28 @@
+data "azurerm_kubernetes_cluster" "current" {
```

Comment:
After applying the node pool configuration for the first time, a second apply wants to replace the node pool. `vnet_subnet_id` change forces a replacement.

```
# module.aks_zero_node_pool.module.node_pool.azurerm_kubernetes_cluster_node_pool.current must be replaced
-/+ resource "azurerm_kubernetes_cluster_node_pool" "current" {
      - availability_zones     = [] -> null
      - enable_host_encryption = false -> null
      - enable_node_public_ip  = false -> null
      - fips_enabled           = false -> null
      ~ id                     = "/subscriptions/803f877a-b0ff-49a7-afc7-eea135679b9a/resourcegroups/terraform-kubestack-testing/providers/Microsoft.ContainerService/managedClusters/kbstacctest-ops-westeurope/agentPools/test1" -> (known after apply)
      ~ kubelet_disk_type      = "OS" -> (known after apply)
        name                   = "test1"
      ~ node_labels            = {} -> (known after apply)
      ~ orchestrator_version   = "1.20.9" -> (known after apply)
      ~ os_disk_size_gb        = 128 -> (known after apply)
      ~ os_sku                 = "Ubuntu" -> (known after apply)
      - tags                   = {} -> null
      - vnet_subnet_id         = "/subscriptions/803f877a-b0ff-49a7-afc7-eea135679b9a/resourceGroups/terraform-kubestack-testing/providers/Microsoft.Network/virtualNetworks/kbstacctest-ops-westeurope/subnets/kbstacctest-ops-westeurope-default-node-pool" -> null # forces replacement
        # (14 unchanged attributes hidden)
    }
```

It's probably best, to make this attribute configurable, but default to the value of the [cluster's data source](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/kubernetes_cluster#vnet_subnet_id).

So something like this in the configuration.tf locals:

```hcl
vnet_subnet_id = local.cfg["vnet_subnet_id"] != null ? local.cfg["vnet_subnet_id"] : data.azurerm_kubernetes_cluster.current.vnet_subnet_id
```

Reply:
One thought, I don't know if it makes sense to have different node pools use different subnet IDs on Azure. It's a thing on AWS. So assuming it does not make sense on Azure, you might as well just set it to what the cluster data source returns.

Reply:
At least on the [terraform side](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster_node_pool#vnet_subnet_id) they have this:
 > At this time the vnet_subnet_id must be the same for all node pools in the cluster

So I'll not expose the config, but rather just take it from the cluster data source

Reply:
Turned out a bit less straightforward, as (at least in my configuration) all the `vnet_subnet_id` values that were returned by the data source were actually empty strings. Seems to be to do the sensible thing, though.

```
❯ terraform state show data.azurerm_kubernetes_cluster.current
# data.azurerm_kubernetes_cluster.current:
data "azurerm_kubernetes_cluster" "current" {
    addon_profile                   = [
 # ...snip
    ]
    agent_pool_profile              = [
        {
            availability_zones       = [
                "1",
                "2",
                "3",
            ]
            count                    = 1
            enable_auto_scaling      = true
            enable_node_public_ip    = false
            max_count                = 5
            max_pods                 = 110
            min_count                = 1
            name                     = "main"
            node_labels              = {
                "component" = "main"
            }
            node_public_ip_prefix_id = ""
            node_taints              = []
            orchestrator_version     = "1.21.2"
            os_disk_size_gb          = 128
            os_type                  = "Linux"
            tags                     = {}
            type                     = "VirtualMachineScaleSets"
            upgrade_settings         = []
            vm_size                  = "Standard_D8ds_v4"
            vnet_subnet_id           = ""
        },
        {
            availability_zones       = []
            count                    = 1
            enable_auto_scaling      = true
            enable_node_public_ip    = false
            max_count                = 2
            max_pods                 = 110
            min_count                = 1
            name                     = "system"
            node_labels              = {}
            node_public_ip_prefix_id = ""
            node_taints              = [
                "CriticalAddonsOnly=true:NoSchedule",
            ]
            orchestrator_version     = "1.21.2"
            os_disk_size_gb          = 30
            os_type                  = "Linux"
            tags                     = {}
            type                     = "VirtualMachineScaleSets"
            upgrade_settings         = []
            vm_size                  = "Standard_B4ms"
            vnet_subnet_id           = ""
        },
        {
            availability_zones       = [
                "1",
                "2",
                "3",
            ]
            count                    = 0
            enable_auto_scaling      = true
            enable_node_public_ip    = false
            max_count                = 3
            max_pods                 = 110
            min_count                = 0
            name                     = "worker"
            node_labels              = {
                "component" = "bi-worker"
            }
            node_public_ip_prefix_id = ""
            node_taints              = []
            orchestrator_version     = "1.21.2"
            os_disk_size_gb          = 128
            os_type                  = "Linux"
            tags                     = {}
            type                     = "VirtualMachineScaleSets"
            upgrade_settings         = []
            vm_size                  = "Standard_D8ds_v4"
            vnet_subnet_id           = ""
        },
    ]
    api_server_authorized_ip_ranges = []
 # ...snip
```